### PR TITLE
[RND-275] Assignable -> SuperclassInfo with simpler semantics

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: Meadowlark-js/backends/meadowlark-aws-dynamodb-opensearch/src/BaseDynamoRepository.ts [hunk 1]
```diff
@@ -120,9 +120,10 @@ export function foreignKeyConditions(resourceInfo: ResourceInfo, documentInfo: D
       isDescriptor: false,
     };
 
-    if (documentReference.isAssignableFrom) {
-      return conditionCheckFromAssignable(entityTypeInfo, documentReference.documentIdentity);
-    }
+    // TODO: Needs updating to RND-275 changes
+    // if (documentReference.isAssignableFrom) {
+    //   return conditionCheckFromAssignable(entityTypeInfo, documentReference.documentIdentity);
+    // }
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-aws-dynamodb-opensearch/src/BaseDynamoRepository.ts:126]: A break of the dynamo code

File: Meadowlark-js/backends/meadowlark-mongodb-backend/src/model/MeadowlarkDocument.ts [hunk 2]
```diff
@@ -47,6 +54,25 @@ export interface MeadowlarkDocument extends MeadowlarkDocumentId {
...
+   * Example: Assessment has a reference to EducationOrganization, and ClassPeriod has a reference to School.
+   *          School is a subclass of EducationOrganization, with schoolId being the identity of School and
+   *          educationOrganizationId being the corresponding identity of EducationOrganization.
+   *
+   *          As a result, a single School document with schoolId=123 will satisfy the validation of both
+   *          a reference from ClassPeriod to a School with schoolId=123, as well as a reference from
+   *          Assessment to EducationOrganization with educationOrganizationId=123.
+   */
+  existenceIds: string[];
+
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-mongodb-backend/src/model/MeadowlarkDocument.ts:75]: Description of the array in MongoDB now used for existence checking for reference validation

File: Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Db.ts [hunk 3]
```diff
@@ -33,6 +33,7 @@ export async function getNewClient(): Promise<MongoClient> {
 
     // Note this does nothing if the index already exists (triggers an index build otherwise)
     await collection.createIndex({ outRefs: 1 });
+    await collection.createIndex({ existenceIds: 1 });
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Db.ts:36]: index for existenceIds

File: Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/ReferenceValidation.ts [hunk 4]
```diff
@@ -5,22 +5,16 @@
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/ReferenceValidation.ts:5]: Mostly a rename of WriteHelper after moving out the one non-reference validation piece

File: Meadowlark-js/packages/meadowlark-core/src/model/DocumentId.ts [hunk 5]
```diff
@@ -44,16 +45,25 @@ export function documentIdForDocumentIdentity(
...
 /**
- * Returns the id of the given DocumentInfo, using the project name, resource name, resource version
- * and identity of the API document.
- * If the given documentInfo is assignable, uses that information for the document id.
+ * Returns the id of the given DocumentInfo, using the project name, resource name and identity of the API document.
  */
 export function documentIdForDocumentInfo(resourceInfo: ResourceInfo, documentInfo: DocumentInfo): string {
-  // If this is an assignable entity, use the assignableIdentity for the id instead of the actual document identity
-  const documentIdentity: DocumentIdentity =
-    documentInfo.assignableInfo == null ? documentInfo.documentIdentity : documentInfo.assignableInfo.assignableIdentity;
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/model/DocumentId.ts:54]: This was convoluted, but removing broke Dynamo's superclass handling.

File: Meadowlark-js/packages/meadowlark-core/src/model/DocumentReference.ts [hunk 6]
```diff
@@ -28,16 +28,6 @@ export type DocumentReference = {
...
-  /**
-   * True if the referenced entity is assignable from other entities (meaning it is a superclass),
-   *
-   * Example 1: School is not a superclass. isAssignableFrom would be false if the reference
-   *            was to a School.
-   * Example 2: EducationOrganization is a superclass of School, LocalEducationAgency, and others. isAssignableFrom
-   *            would be true if the reference was to an EducationOrganization.
-   */
-  isAssignableFrom: boolean;
-
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/model/DocumentReference.ts:40]: Unnecessary due to separation of documentId from superclass documentId

File: Meadowlark-js/packages/meadowlark-core/src/model/SuperclassInfo.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+import { DocumentIdentity } from './DocumentIdentity';
+
+/**
+ * The optional superclass information for a DocumentInfo. Applies only to documents that are subclasses,
+ * providing superclass identity information. (Note the ODS/API allows only a single level of subclassing.)
+ */
+export type SuperclassInfo = {
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/model/SuperclassInfo.ts:12]: A rename of Assignable, and alignment of equivalent property names to be like those in DocumentInfo

File: Meadowlark-js/packages/meadowlark-core/src/validation/DocumentReferenceExtractor.ts [hunk 8]
```diff
@@ -186,8 +180,7 @@ function documentReferencesFromReferenceGroup(
       projectName: referenceGroup.sourceProperty.namespace.projectName,
       resourceName,
       resourceVersion: referenceGroup.sourceProperty.namespace.projectVersion,
-      documentIdentity: assignable == null ? documentIdentity : assignable.assignableIdentity,
-      isAssignableFrom: referencedEntity.subclassedBy.length > 0,
+      documentIdentity,
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/validation/DocumentReferenceExtractor.ts:183]: simplified

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiEntityMappingEnhancerBase.ts [hunk 9]
```diff
@@ -51,14 +51,13 @@ export function descriptorCollectedPropertiesFrom(entity: TopLevelEntity): Colle
...
  * which it can be assigned to, otherwise null. Note MetaEd only allows a single level of subclassing.
  */
-export function assignabilityFor(entity: TopLevelEntity): TopLevelEntity | null {
+export function superclassFor(entity: TopLevelEntity): TopLevelEntity | null {
   // Specifically exclude Domain Entity/Association Extensions - just to be safe
   if (entity.type === 'domainEntityExtension' || entity.type === 'associationExtension') return null;
   // If it's a subclass, return its superclass
   if (entity.baseEntity != null) return entity.baseEntity;
-  // If it's a superclass, return itself
-  return entity.subclassedBy.length > 0 ? entity : null;
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiEntityMappingEnhancerBase.ts:63]: Another simpilfication, assignable used to need to exist on the superclass, not just the subclasses.

File: Meadowlark-js/tests/system/crud/SystemTestSetup.ts [hunk 10]
```diff
@@ -62,15 +62,15 @@ export function schoolGetClient1(): FrontendRequest {
   return {
     ...newFrontendRequestTemplate(),
     headers: { ...JSON_HEADER, ...CLIENT1_HEADERS },
-    path: '/v3.3b/ed-fi/schools/x1GptgyYapmpBGiZegIbM2XC_NLMVVjisNLEtg',
+    path: '/v3.3b/ed-fi/schools/L9gXuk9vioIoG64QKp8NFO2f3AOe78fV-HrtfQ',
```
Comment (bradbanister) [inline Meadowlark-js/tests/system/crud/SystemTestSetup.ts:65]: Actually a bug fix, see below

File: Meadowlark-js/tests/system/crud/Upsert.test.ts [hunk 11]
```diff
@@ -158,7 +158,7 @@ describe('given a POST of an academic week referencing a school that does not ex
 
   it('should return failure due to missing reference', async () => {
     expect(upsertResult.body).toMatchInlineSnapshot(
-      `"{\\"message\\":\\"Reference validation failed: Resource School is missing identity [{\\\\\\"name\\\\\\":\\\\\\"educationOrganizationId\\\\\\",\\\\\\"value\\\\\\":123}]\\"}"`,
+      `"{\\"message\\":\\"Reference validation failed: Resource School is missing identity [{\\\\\\"name\\\\\\":\\\\\\"schoolId\\\\\\",\\\\\\"value\\\\\\":123}]\\"}"`,
```
Comment (bradbanister) [inline Meadowlark-js/tests/system/crud/Upsert.test.ts:161]: educationOrganizationId was a bug

File: Meadowlark-js/backends/meadowlark-aws-dynamodb-opensearch/src/BaseDynamoRepository.ts [hunk 12]
```diff
@@ -229,13 +230,13 @@ export function constructAssignableDeleteItem(
   resourceInfo: ResourceInfo,
   documentInfo: DocumentInfo,
 ): TransactWriteItem | null {
-  if (documentInfo.assignableInfo == null) return null;
+  if (documentInfo.superclassInfo == null) return null;
 
-  // TODO: Note for the future, this assumes the "assignable to" entity is in the same project/namespace as the given entity
+  // TODO: Note for the future, this assumes the "superclass" entity is in the same project/namespace as the given entity
```
Comment (stephenfuqua) [inline Meadowlark-js/backends/meadowlark-aws-dynamodb-opensearch/src/BaseDynamoRepository.ts:235]: I just started https://techdocs.ed-fi.org/display/EFTD/Meadowlark+-+Miscellaneous+Notes to help remember issues like this. I'm not ready to fill up the Jira backlog just yet. Don't want to get too far ahead of ourselves.

File: Meadowlark-js/backends/meadowlark-aws-dynamodb-opensearch/src/BaseDynamoRepository.ts [hunk 13]
```diff
@@ -120,9 +120,10 @@ export function foreignKeyConditions(resourceInfo: ResourceInfo, documentInfo: D
       isDescriptor: false,
     };
 
-    if (documentReference.isAssignableFrom) {
-      return conditionCheckFromAssignable(entityTypeInfo, documentReference.documentIdentity);
-    }
+    // TODO: Needs updating to RND-275 changes
```
Comment (stephenfuqua) [inline Meadowlark-js/backends/meadowlark-aws-dynamodb-opensearch/src/BaseDynamoRepository.ts:123]: Does that mean it _will_ be done in RND-275? If not, please create a separate ticket. If we axe the Dynamo support then we can just close that ticket.

File: Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Db.ts (hunk 3 above)
Comment (stephenfuqua) [reply Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Db.ts:36]: This caused me to ask: what are the limits on indexes? [We're in safe territory](https://www.mongodb.com/docs/manual/reference/limits/#indexes), not surprisingly.

File: Meadowlark-js/backends/meadowlark-mongodb-backend/src/model/MeadowlarkDocument.ts [hunk 14]
```diff
@@ -47,6 +54,25 @@ export interface MeadowlarkDocument extends MeadowlarkDocumentId {
...
+   *
+   * Example: Assessment has a reference to EducationOrganization, and ClassPeriod has a reference to School.
+   *          School is a subclass of EducationOrganization, with schoolId being the identity of School and
+   *          educationOrganizationId being the corresponding identity of EducationOrganization.
+   *
+   *          As a result, a single School document with schoolId=123 will satisfy the validation of both
+   *          a reference from ClassPeriod to a School with schoolId=123, as well as a reference from
+   *          Assessment to EducationOrganization with educationOrganizationId=123.
+   */
+  existenceIds: string[];
```
Comment (stephenfuqua) [inline Meadowlark-js/backends/meadowlark-mongodb-backend/src/model/MeadowlarkDocument.ts:74]: I can live with `existenceIds`, but anyone not reading this (very helpful) comment may have a hard time grasping the name when seeing it elsewhere. 

So far, the only other word I can come up with is "aliases".

File: Meadowlark-js/packages/meadowlark-core/src/model/DocumentId.ts (hunk 5 above)
Comment (stephenfuqua) [reply Meadowlark-js/packages/meadowlark-core/src/model/DocumentId.ts:54]: So... we end up with a function that just delegates to another function? Can we cut out the middleman? Or would it make sense to move this function into `DocumentInfo`?

File: Meadowlark-js/packages/meadowlark-core/src/model/DocumentId.ts [hunk 15]
```diff
@@ -44,16 +45,25 @@ export function documentIdForDocumentIdentity(
...
+ * Returns the id of the given DocumentInfo superclass, using the project name, resource name
+ * and identity of the superclass document.
+ */
+export function documentIdForSuperclassInfo(superclassInfo: SuperclassInfo): string {
+  const resourceInfo: ResourceInfo = {
+    projectName: superclassInfo.projectName,
+    resourceName: superclassInfo.resourceName,
+    resourceVersion: '', // Not relevant
+    isDescriptor: false, // Descriptors are never superclasses
+  };
```
[the remark this answers is not in the document]
Comment (stephenfuqua) [reply Meadowlark-js/packages/meadowlark-core/src/model/DocumentId.ts:64]: Likewise, could we move this to `SuperclassInfo`?

File: Meadowlark-js/packages/meadowlark-core/src/model/DocumentId.ts (hunk 5 above)
Comment (bradbanister) [reply Meadowlark-js/packages/meadowlark-core/src/model/DocumentId.ts:54]: 👍

File: Meadowlark-js/packages/meadowlark-core/src/model/DocumentId.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (bradbanister) [reply Meadowlark-js/packages/meadowlark-core/src/model/DocumentId.ts:63]: 👍

**Review by stephenfuqua — APPROVED** [review]

**Pull request merged**